### PR TITLE
change(web): adds error + console logs usable for Sentry reporting targeting #4797

### PR DESCRIPTION
**Pull request by jahorton**
Sadly, I'm currently unable to reproduce #4797, despite the fact that we're usually seeing at least a couple counts of it per day from keymanweb.com's Sentry logs.  So, to better assist diagnosis of the cause, this PR adds 'breadcrumbs' that will provide extra information about the error's context.

While `target.innerHTML` (or `target.parent.innerHTML`, or the like) might seem to be a more simple, straightforward solution... that would also include the user's typed input, which is obviously a privacy concern.  Rather than attempt complicated regex to scrub it out at all levels... I instead opted to manually collate relevant info using the current approach.

Will need to be 🍒-picked to stable in order for keymanweb.com to provide us anything useful, of course.

Example resulting log:  https://sentry.keyman.com/organizations/keyman/issues/4658/?project=10&query=is%3Aunresolved

### CODE REVIEW

**Comment by mcdurdin** [comment]
This change is a bit too complex for me to feel comfortable approving in time to get in a PR for stable-14.0 today. Will continue to review and discuss with you.

**Comment by jahorton** [comment]
> This change is a bit too complex for me to feel comfortable approving in time to get in a PR for stable-14.0 today. Will continue to review and discuss with you.

I suppose, given the team plans for the next two weeks.  That said... this _is_ why the PR went up a couple days ago.

**Comment by mcdurdin** [comment]
> I suppose, given the team plans for the next two weeks. That said... this _is_ why the PR went up a couple days ago.

Well, I started reviewing a couple of days ago ... but I'm still not comfortable with the changes! Too many other things I am having to do (videos, support, etc).

File: web/source/dom/domEventHandlers.ts [hunk 1]
```diff
@@ -563,8 +566,9 @@ namespace com.keyman.dom {
 
       // Determines the actual TouchAliasElement - the part tied to an OutputTarget.
       // Ideally, we shouldn't need the second part as a fallback; it's there to preserve existing
-      // behavior from 13.0, as this was refactored made LATE in the 14.0 beta process.
-      let target = findTouchAliasTarget(tTarg) || (tTarg as TouchAliasElement);
+      // behavior from 13.0, as this was refactored LATE in the 14.0 beta process.
+      let target = findTouchAliasTarget(tTarg);
```
Comment (jahorton) [inline web/source/dom/domEventHandlers.ts:570]: The current logs we see for #4797 were likely from here; failure to find a `TouchAliasElement` from `tTarg` would generally lead to `setTextCaret` being `undefined` as seen in the logs.

So, this _removes_ that fallback.

File: web/source/dom/domEventHandlers.ts [hunk 2]
```diff
@@ -206,7 +206,9 @@ namespace com.keyman.dom {
 
       // Makes sure we properly detect the TouchAliasElement root, 
       // rather than one of its constituent children.
-      Ltarg = findTouchAliasTarget(Ltarg) || Ltarg;
+      if(this.keyman.util.device.touchable) {
+        Ltarg = findTouchAliasTarget(Ltarg);
+      }
```
Comment (jahorton) [inline web/source/dom/domEventHandlers.ts:211]: While this block doesn't seem to be affected... _currently_... it makes the same assumptions that lead to the error we do see.

File: web/source/dom/domEventHandlers.ts [hunk 3]
```diff
@@ -791,10 +801,8 @@ namespace com.keyman.dom {
       }
       
       // Identify the input element from the touch event target (touched element may be contained by input)
-      if(target.className == null || target.className.indexOf('keymanweb-input') < 0) target=<HTMLElement> target.parentNode;
-      if(target.className == null || target.className.indexOf('keymanweb-input') < 0) target=<HTMLElement> target.parentNode;
-      if(target.className == null || target.className.indexOf('keymanweb-input') < 0) return;
-      
+      target = findTouchAliasTarget(target);
+
```
Comment (jahorton) [inline web/source/dom/domEventHandlers.ts:805]: The original lines did what the new line does.  I missed that in #4638.

File: web/source/dom/touchAliasElement.ts [hunk 4]
```diff
@@ -30,26 +30,83 @@ namespace com.keyman.dom {
...
+        console.info(`Next sibling: ${elementDescriber(target.nextSibling)}`);
+      }
+      console.info(`Target's parent: ${elementDescriber(target.parentElement)}`);
+      if(target.parentElement instanceof HTMLElement) {
+        console.info(`Target's grandparent: ${elementDescriber(target.parentElement.parentElement)}`);
+      }
+      // Triggers the actual Sentry log (on pages that include Sentry logging)
+      throw new Error(`Unexpected touch-start event target.`);
+    };
+
```
Comment (jahorton) [inline web/source/dom/touchAliasElement.ts:82]: end:  definition of the logging function.

File: web/source/dom/touchAliasElement.ts [hunk 5]
```diff
@@ -30,26 +30,83 @@ namespace com.keyman.dom {
...
     } else if(target['kmw_ip']) { // In case it's called on a TouchAliasElement's base (aliased) element.
       return target['kmw_ip'] as TouchAliasElement;
     } else {
       // If it's not in any way related to a TouchAliasElement, simply return null.
+      // NOTE:  At present, this method should only be called in places that actually EXPECT a TouchAliasElement...
+      // So if we fail to find one, something's gone wrong, and we'd like more details.
+
+      // There's an evasive error related to this at present so we'll temporarily console-log info 
+      // that may help us determine why we couldn't find the expected element.
+      elementLogger();
```
Comment (jahorton) [inline web/source/dom/touchAliasElement.ts:109]: So, where before we were returning `null` for the unexpected 'fallback' case (which, again, leads to the `setTextCaret` - `undefined` error), we're now logging + throwing errors to trigger Sentry log capture on keymanweb.com with something more informative.

**Review by jahorton — COMMENTED** [review]
Added some explanation to the changeset's most significant points.

File: web/source/dom/domEventHandlers.ts [hunk 6]
```diff
@@ -602,8 +606,9 @@ namespace com.keyman.dom {
         osk._Show();
       }
       
-      // If clicked on DIV on the main element or on the scroller element, set caret to end of text
-      if(tTarg && tTarg == target || tTarg == scroller) {
+      // If clicked on DIV on the main element, rather than any part of the text representation,
+      // set caret to end of text
+      if(tTarg && tTarg == target) {
```
Comment (mcdurdin) [inline web/source/dom/domEventHandlers.ts:611]: What if `tTarg == scroller` now? Can that happen? (The previous condition was horrible with its mix of && and || but I'm not clear on the change here.)

File: web/source/dom/domEventHandlers.ts [hunk 7]
```diff
@@ -614,8 +619,9 @@ namespace com.keyman.dom {
         }
     
         target.setTextCaret(cp);
-        target.scrollInput();        
-      } else { // Otherwise, if clicked on text in SPAN, set at touch position
+        target.scrollInput();
+        // nextSibling - the scrollbar element.  
+      } else if(tTarg != scroller.nextSibling) { // Otherwise, if clicked on text in SPAN, set at touch position
```
Comment (mcdurdin) [inline web/source/dom/domEventHandlers.ts:624]: What is `scroller.nextSibling` -- why this change?

File: web/source/dom/domEventHandlers.ts [hunk 8]
```diff
@@ -774,8 +780,12 @@ namespace com.keyman.dom {
      * Handle the touch move event for an input element
      */         
     dragInput: (e: TouchEvent|MouseEvent) => void = function(this: DOMTouchHandlers, e: TouchEvent|MouseEvent) {
-      // Prevent dragging window 
-      e.preventDefault();
+      // Prevent dragging window
+      if(e.cancelable) {
```
Comment (mcdurdin) [inline web/source/dom/domEventHandlers.ts:784]: I assume this fix is unrelated to the primary fix and diagnostic log?

File: web/source/dom/domEventHandlers.ts (hunk 1 above)
Comment (mcdurdin) [reply web/source/dom/domEventHandlers.ts:570]: Doesn't this change make the comment above incorrect?

**Review by mcdurdin — APPROVED** [review]
I'm a little uncomfortable with fixes and diagnostic logging in the same PR -- ideally they'd be separated (even if chained). But I think it's okay, once we address the comments I have made. The comments you added to the PR certainly helped 👍 .

Can you make a cherry-pick to stable please?

File: web/source/dom/touchAliasElement.ts [hunk 9]
```diff
@@ -30,26 +30,83 @@ namespace com.keyman.dom {
...
+      }
+      if(target.nextSibling && target.nextSibling instanceof HTMLElement) {
+        console.info(`Next sibling: ${elementDescriber(target.nextSibling)}`);
+      }
+      console.info(`Target's parent: ${elementDescriber(target.parentElement)}`);
+      if(target.parentElement instanceof HTMLElement) {
+        console.info(`Target's grandparent: ${elementDescriber(target.parentElement.parentElement)}`);
+      }
+      // Triggers the actual Sentry log (on pages that include Sentry logging)
+      throw new Error(`Unexpected touch-start event target.`);
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/dom/touchAliasElement.ts:80]: Web does not currently have _any_ baked-in Sentry access - not even the 'minimal' version.  That function is therefore unaccessible.

File: web/source/dom/domEventHandlers.ts (hunk 1 above)
Comment (jahorton) [reply web/source/dom/domEventHandlers.ts:570]: Right, I should remove the

> Ideally, ...

part when this goes through.

File: web/source/dom/domEventHandlers.ts (hunk 7 above)
Comment (jahorton) [reply web/source/dom/domEventHandlers.ts:624]: `scroller.nextSibling` - the simulated _scrollbar_.

Not the scrolling `div` used to scroll the element's text contents - the scrollbar element used to control the scrolling position.  Which is therefore _not_ a click within the bounds of the user's text.

File: web/source/dom/domEventHandlers.ts (hunk 6 above)
Comment (jahorton) [reply web/source/dom/domEventHandlers.ts:611]: If `tTarg == scroller`, that means the click was within the text-containing area.  So, we handle that in the next 'rung' of the if-else sequence - the condition of `if(tTarg != scroller.nextSibling)` will be `true` in that case.

File: web/source/dom/domEventHandlers.ts (hunk 8 above)
Comment (jahorton) [reply web/source/dom/domEventHandlers.ts:784]: This was generating extraneous error messages that we don't need to capture.  It mostly shows up during development testing, at least in my experience.

When it does show, it gets _very_ noisy.  The following is taken from (what a user would perceive as) a simple, single scroll.  (Took several attempts to repro, but when it _does_ trigger...)

Screen Shot 2021-04-19 at 8 29 44 AM

Note that Chrome's error text tells me rather directly to make this adjustment.

File: web/source/dom/touchAliasElement.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/dom/touchAliasElement.ts:80]: > This keeps the existing code path, even if it is currently broken, rather than introducing a new exception code path.

They're the same path.  We simply report the error at a more informative location.

File: web/source/dom/domEventHandlers.ts (hunk 6 above)
Comment (mcdurdin) [reply web/source/dom/domEventHandlers.ts:611]: Sure. But that means if `tTarg == scroller`, we are now running an entirely different piece of code which seems to be doing some quite different things, so not quite sure how that works out?

Comment (jahorton) [reply web/source/dom/domEventHandlers.ts:611]: Right.  That's a minor fix from the change made in https://github.com/keymanapp/keyman/pull/4638, as I actually shouldn't have included `scroller` in the check originally.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.32-alpha](https://keyman.com/downloads/releases/alpha/15.0.32)
